lib/mapObjects/CRewardableObject.h

```diff
@@ -70,13 +70,6 @@ class DLL_LINKAGE CRewardLimiter
 	/// sub-limiters, none should pass for this limiter to pass
 	TRewardLimitersList noneOf;
 
-	CRewardLimiter():
-		dayOfWeek(0),
-		daysPassed(0),
-		heroLevel(0),
-		primary(4, 0)
-	{}
-
```

Comment:
Was this hard?
I am tired of this arguing and of searching for new regressions.

```
	CRewardLimiter():
		dayOfWeek(0),
		daysPassed(0),
		heroExperience(0),
		heroLevel(0),
		manaPercentage(0),
		manaPoints(0),
		primary(GameConstants::PRIMARY_SKILLS, 0)
	{}
```

Reply:
Okay, I will do this, this code will be reworked anyway after merging of banks and rewardables (in 1.3).

Reply:
Done.

You simply dislike default initialization at all and prefer to write constructors everywhere. Understood, will not touch this in any old code.

Reply:
I guess I am...

Was ambivalent a month before, but was completely unimpressed by how it turned out in vcmi code...

Reply:
I just think than in some rare cases it is good (like in structs which lives entirely on header). For example, here we have clearly 3 structs (and two is almost the same, so, if I will do rewardables after moats, I think merging RewardInfo and RewardLimiter is reasonable). For CCreature, Bonus, and other big structs with multiple constructors - yes, it is reasonable to init everything in a constructor. Anyway, by reason you said in Slack, VCMI have very bad code even by C++11 standard.

Reply:
I would say there is a good engineering practice. Do not touch what works). So this is more about weird instructions which does not always work. C# for instance also introduced inline initialization as a new cool feature and we use it now at work. But nobody change old code just for fun). Only if you touch some file it is allowed to make mistakeless automatic refactoring. Why in C++ it can break just because you replaced on logically equivalent thing with another? Maybe there is something we do not know behind it? What is the order of default initialization and constructor if both defined?

Reply:
@nullkiller there is interesting property for structs without a constructor: https://en.cppreference.com/w/cpp/language/aggregate_initialization.

If we define a constructor, we cannot use aggregate initialization. Default values allow us to specify default values even without an explicit constructor. 

I think for classes that can become aggregates with removal of one default constructor - usage of default members is reasonable.

Reply:
We use c++14, so, we can use default member initializers on aggregates)

Reply:
@IvanSavenko @nullkiller I mainly for converting this classes:
```c++
// aggregate
struct base1 { int b1, b2 = 42; };
 
// non-aggregate
struct base2
{
    base2() : b3(42) {}
 
    int b3;
};
 ```

Reply:
> Why in C++ it can break just because you replaced on logically equivalent thing with another?

It does not. All regressions were from incorrect initialization. For example, initializer was removed from constructor but not added to member declaration. So far with all "modernizations" I found this issue 3 times. Hopefully that's all of them, but I can only hope.

Also, this is personal opinion, but I dislike how it turns out in code, For example our typical library class right now looks like this:
```
class CArtifact
{
	ArtifactID id;
	std::string identifier;
	EartClass aClass = ART_SPECIAL;
	ui32 price = 0;
}
```
1) `id` field is not initialized in place - this may be effectively enum/integer, but since it is wrapped in class, ArtifactID class will perform its initialization instead.
2) `identifier` is not initialized in place either - since this is also a class, which handles its own initialization
3) `aClass` however uses in-place initialization. Because unlike `id` this is enum which has to be explicitly initialized with valid enum value;
4) `price` also uses in-place initialization since this is also primitive type.

As result we end up with this mess where half of members use in-place init, but half members do not. Before using in-place initialization this was not an issue. And these default values often don't make sense.

Another issue is readability - this change increases amount of text in headers, making it harder to get a quick look on class API. Not a problem for smaller classes that are 10 lines in size, but when you have class with definition that is 100-200 lines long I would rather keep as much info as possible external.

Reply:
@rilian-la-te 
If not for our serializer that requires default constructor I would have proposed to get rid of both default constructors and this aggregate initialization and instead give all such small structs constructor that *always* initializes all members, e.g.
```
class Point {
    int x, y;
    Point(int x, int y) : x(x), y(y) {}
}
```
Such form would have allowed us to
1) Always initialize all class members with *meaningful* values instead of zero-init
2) Allow us to detect/update code that does not provides (meaningful) values to some of the fields.

It is already a pain to find all instances in code where certain class (e.g. netpack) is created and update it, e.g. if you add new member. Aggregate init would make it even worse.

Reply:
@IvanSavenko I disagree with you about readability in some cases. When we have constructor in header (or entire class in *.cpp, readability without constructor is better). About second proposal - it is pain in the ass to create a temporary values with deleted default constructor.